app/src/main/java/com/github/se/polyfit/data/remote/firebase/PostFirebaseRepository.kt

```diff
@@ -41,4 +61,110 @@ class PostFirebaseRepository(
+   *   posts.
+   */
+  fun queryNearbyPosts(
+      centerLatitude: Double,
+      centerLongitude: Double,
+      radiusInKm: Double,
+      geoFire: GeoFire = GeoFire(geoFireRef)
+  ) {
+    val center = GeoLocation(centerLatitude, centerLongitude)
+    val query = geoFire.queryAtLocation(center, radiusInKm)
+
+    query.addGeoQueryEventListener(
+        object : GeoQueryEventListener {
+          val nearbyKeys = mutableListOf<String>()
+
+          override fun onKeyEntered(key: String, location: GeoLocation) {
+            nearbyKeys.add(key)
+          }
+
+          override fun onKeyExited(key: String) {
+            nearbyKeys.remove(key)
+          }
+
+          override fun onKeyMoved(key: String, location: GeoLocation) {
+            // Do nothing but need to be there
+          }
+
+          override fun onGeoQueryReady() {
+            PostFirebaseRepository().fetchPostsAndImages(nearbyKeys) { posts ->
+              // Do something with the fetched posts
+            }
+          }
+
+          override fun onGeoQueryError(error: DatabaseError) {
+            Log.e("GeoQuery", "There was an error with the geo query: ${error.message}")
+          }
+        })
+  }
+
+  fun fetchPostsAndImages(
+      keys: List<String>,
+      postCollection: CollectionReference = this.postCollection,
+      completion: (List<Post>) -> Unit
+  ) {
+    val posts = mutableListOf<Post>()
+    val batchSize = 10
+    val batches = keys.chunked(batchSize)
+    var completedBatches = 0
+
+    // Use a batch operation to fetch all posts
+    batches.forEach { batch ->
+      postCollection
+          .whereIn(FieldPath.documentId(), batch)
+          .get()
+          .addOnSuccessListener { querySnapshot ->
+            CoroutineScope(Dispatchers.IO).launch {
+              val tempPosts = mutableListOf<Post>()
+              val deferredImages = mutableListOf<Deferred<Unit>>()
+
+              querySnapshot.documents.forEach { document ->
+                document.data?.let {
+                  Post.deserialize(it)?.also { post ->
+                    deferredImages.add(
+                        async { post.listOfURLs = fetchImageReferencesForPost(document.id) })
+                    tempPosts.add(post)
+                  }
+                }
+              }
+              deferredImages.awaitAll()
+
+              synchronized(posts) {
+                posts.addAll(tempPosts)
+                completedBatches++
+                if (completedBatches == batches.size) {
+                  completion(posts)
+                }
+              }
+            }
+          }
+          .addOnFailureListener {
+            Log.e("PostFirebaseRepository", "Failed to fetch posts: ${it.message}")
+          }
+    }
+  }
+
+  suspend fun fetchImageReferencesForPost(postKey: String): List<StorageReference> {
+    return withContext(Dispatchers.IO) {
+      val storageRef = pictureDb.getReference("posts/$postKey")
```

Comment:
Maybe 'posts/' could be stored in a string resource file to avoid hardcoded references.